src/views/workflow-history/workflow-history.types.ts

```diff
@@ -0,0 +1,160 @@
+import type { HistoryEvent } from '@/__generated__/proto-ts/uber/cadence/api/v1/HistoryEvent';
+
+import type { WorkflowEventStatus } from './workflow-history-event-status-badge/workflow-history-event-status-badge.types';
+
+export type HistoryEventGroupType =
+  | 'Activity'
+  | 'Decision'
+  | 'Timer'
+  | 'ChildWorkflowExecution'
+  | 'SignalExternalWorkflowExecution'
+  | 'RequestCancelExternalWorkflowExecution'
+  | 'Event';
+
+export type HistoryGroupEventMetadata = {
+  label: string;
+  status: WorkflowEventStatus;
+  timeMs: number | null;
+  timeLabel: string;
+};
+
+export type HistoryGroupEventToStatusMap<GroupT extends HistoryEventsGroup> =
+  Record<
+    GroupT['events'][number]['attributes'],
+    | WorkflowEventStatus
+    | ((
+        event: GroupT['events'][number],
+        events: GroupT['events'],
+        index: number
+      ) => WorkflowEventStatus)
+  >;
+
+export type HistoryGroupEventToStringMap<GroupT extends HistoryEventsGroup> =
+  Record<GroupT['events'][number]['attributes'], string>;
+
+type BaseHistoryGroup = {
+  label: string;
+  eventsMetadata: HistoryGroupEventMetadata[];
+  status: WorkflowEventStatus;
+  hasMissingEvents: boolean;
+  timeMs: number | null;
+  timeLabel: string;
+};
+
+export type ActivityHistoryGroup = BaseHistoryGroup & {
+  groupType: 'Activity';
+  events: ActivityHistoryEvent[];
+};
+
+export type DecisionHistoryGroup = BaseHistoryGroup & {
+  groupType: 'Decision';
+  events: DecisionHistoryEvent[];
+};
+
+export type TimerHistoryGroup = BaseHistoryGroup & {
+  groupType: 'Timer';
+  events: TimerHistoryEvent[];
+};
+
+export type ChildWorkflowExecutionHistoryGroup = BaseHistoryGroup & {
+  groupType: 'ChildWorkflowExecution';
+  events: ChildWorkflowExecutionHistoryEvent[];
+};
+
+export type SignalExternalWorkflowExecutionHistoryGroup = BaseHistoryGroup & {
+  groupType: 'SignalExternalWorkflowExecution';
+  events: SignalExternalWorkflowExecutionHistoryEvent[];
+};
+
+export type RequestCancelExternalWorkflowExecutionHistoryGroup =
+  BaseHistoryGroup & {
+    groupType: 'RequestCancelExternalWorkflowExecution';
+    events: RequestCancelExternalWorkflowExecutionHistoryEvent[];
+  };
+
+export type SingleEventHistoryGroup = BaseHistoryGroup & {
+  groupType: 'Event';
+  events: SingleHistoryEvent[];
+};
+
```

Comment:
If you want a generic HistoryEventAttribute type, I unfortunately don't see one in the code, but you can try something like `type HistoryEventAttribute = HistoryEvent['attributes']`